### PR TITLE
Fix RenameCells.Seurat to support multi-assay inputs

### PR DESCRIPTION
**Pull request by longmanz**
With the help from @dcollins15 , this PR fixes:
- https://github.com/satijalab/seurat-object/issues/230

Previously, when users try to run `RenameCells()` on a multi-assay object, it will throw an error because RenameCells mistakenly uses `Cells()` to retrieve the cell IDs in the object, but `Cells()` only returns the cell IDs in the default assay but not the whole object. Changing `Cells()` into `colnames()` now fixes the issue. 

In addition to that, I also added some `if` checks to the function. The new `RenameCells()` now lets users provide a named vector to `new.names = `, so that it can run RenameCells() with an unordered mapping for the cell names.

I have also added a few tests to `test_objects.R` to test the `RenameCells()`.

### CODE REVIEW

File: tests/testthat/test_objects.R [hunk 1]
```diff
@@ -357,3 +357,23 @@ test_that("Update keys works", {
...
+    test_case2 = RenameCells(test_case, new.names = paste0("Cell_", 1:80))
+    expect_equal(colnames(test_case2)[1:3], c("Cell_1", "Cell_2", "Cell_3"))
+    # 
+    new_names = paste0("Cell_", 1:80)
+    names(new_names) = colnames(test_case)[c(41:80, 1:40)]
+    test_case2 = RenameCells(test_case, new.names = new_names)
+    expect_equal(colnames(test_case2)[1:3], c("Cell_41", "Cell_42", "Cell_43"))
+})
+
+
```
Comment (dcollins15) [inline tests/testthat/test_objects.R:379]: 😍 Tests!! Hell yes!! 😍 These do an excellent job at catching the relevant edge case 🤘 

I think it's probably worth starting a new `test_seurat.R` module so we can start fresh and better isolate our tests from one another. 

It will certainly need to be refined over time, but we can introduce a simple test fixture so that we don't have to rely on `pbmc_small` for tests where we don't need real data. 

```R
#' Returns a random counts matrix.
get_random_counts <- function(ncells, nfeatures) {
  # Populate a `nfeatures` x `ncells` matrix with zeros.
  counts <- matrix(0, ncol = ncells, nrow = nfeatures)

  # Assign column and row names to the matrix to label cells and genes.
  colnames(counts) <- paste0("cell", seq(ncol(counts)))
  row.names(counts) <- paste0("gene", seq(nrow(counts)))

  # Convert `counts` to a `dgCMatrix`.
  counts_sparse <- as.sparse(counts)

  return(counts_sparse)
}
```

With that in place, we can extend your test and keep it isolated from any others that we add later (since `test_case` is limited to the scope of just this one test). 

```R
context("RenameCells")

test_that("`RenameCells` works with multi-assay inputs", {
  counts <- get_random_counts(ncells = 10, nfeatures = 10)

  # Run checks against `Assay` and `Assay5` instances.
  for (assay_version in c("v3", "v5")) {
    create_assay <- switch(assay_version,
      v3 = CreateAssayObject,
      v5 = CreateAssay5Object,
      stop("`assay_version` should be one of 'v3', 'v5'")
    )

    counts_left <- counts[, 1:5]
    assay_left <- create_assay(counts_left)
    left <- CreateSeuratObject(assay_left)

    counts_right <- counts[, 6:10]
    assay_right <- create_assay(counts_right)
    right <- CreateSeuratObject(assay_right, assay = "right")

    test_case <- merge(left, right)
    old_names <- colnames(test_case)

    # Rename all the cells in the input.
    new_names <- paste0("new-cell", 1:10)
    result <- RenameCells(test_case, new.names = new_names)
    expect_identical(colnames(result), new_names)

    # Rename just the cells in the input's default assay.
    new_names <- paste0("new-cell", 1:5)
    expected_names <- c(new_names, tail(old_names, n = 5))
    result <- RenameCells(test_case, new.names = new_names)
    expect_identical(colnames(result), expected_names)

    # Rename a subset of cells across both assays.
    new_names <- paste0("new-cell", 2:9)
    expected_names <- c(head(old_names, n = 1), new_names, tail(old_names, n = 1))
    # Requires a named vector to be passed to `new.names`.
    expect_error(RenameCells(test_case, new.names = new_names))
    names(new_names) <- old_names[2:9]
    result <- RenameCells(test_case, new.names = new_names)
    expect_identical(colnames(result), expected_names)

    # All cells provided in a named vector mapping must exist.
    new_names <- paste0("new-cell", 1:10)
    names(new_names) <- c("not-a-cell", tail(old_names, n = 1))
    expect_error(RenameCells(test_case, new.names = new_names))
  }
})

```

File: R/assay5.R [hunk 2]
```diff
@@ -1545,19 +1545,23 @@ VariableFeatures.StdAssay <- function(
...
+      }
+      if (isTRUE(x = simplify) & (is.null(x = layer) || any(is.na(x = layer))) &
+          (is.infinite(x = nfeatures) || length(x = var.features) <=
+           nfeatures)) {
+          return(var.features) 
+      } else if (isTRUE(x = simplify) & (is.null(x = layer) || any(is.na(x = layer))) &
+                 (is.infinite(x = nfeatures) || length(x = var.features) >
+                  nfeatures)) {
+          return(var.features[1:nfeatures])
+      }
```
Comment (dcollins15) [inline R/assay5.R:1564]: We should rebase `longmanz:fix_RenameCells` onto `main` and drop bb5b6e7b75fdada7999cf6867e0c2889b0290a72 so that we can merge it separately from:
-  https://github.com/satijalab/seurat-object/pull/236

**Review by dcollins15 — CHANGES_REQUESTED** [review]
Thanks for this one @longmanz!! Looking great so far 🙌 

I've added a few comments that tweak the behavior of `new.names` to allow a subset of cell names to be provided if a named vector is passed in. 

Keep up the great work! 🚀

**Comment by dcollins15** [comment]
@longmanz per yesterday's discussion, I'm going to go ahead and finish this one up. I'll start by rebasing back onto `main` and then push up my changes 🚀
